compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp

```diff
@@ -405,19 +413,114 @@ CollapseInfo::getTransformedCollapsableLoops(AffineMap map) const {
+  if (failed(consumerToProducerMap)) {
+    return this->clear();
+  }
+
+  FailureOr<CollapsableLoopsSet> consumerCollapsable =
+      consumerInfo.getTransformedCollapsableLoops(
+          consumerToProducerMap.value());
+  if (failed(consumerCollapsable)) {
+    return clear();
+  }
+
+  FailureOr<SmallVector<ReassociationIndices>> consumerReassoc =
+      consumerInfo.getTransformedReassociation(consumerToProducerMap.value());
+  assert(succeeded(consumerReassoc));
+
+  // Get a map from original index to the index it gets collapsed into
+  llvm::DenseMap<long, long> consumerCollapseMap;
+  for (const auto &[idx, indicies] : llvm::enumerate(consumerReassoc.value())) {
+    for (const auto elem : indicies) {
+      consumerCollapseMap[elem] = idx;
+    }
+  }
+
+  // Remove all collapsable loops in `producer` that are not collapsable in
+  // `consumer` (set intersect)
+  bool didChange = collapsableLoops.remove_if([&](long elem) -> bool {
+    return !consumerCollapsable.value().contains(elem);
+  });
+
+  // Now update the reassociation indicies given the updated `collapsableLoops`
+  // and `consumerCollapsableMap`.
+  // The idea is to reconstruct the reassociation indicies, and at each index:
+  // (1) If `index` IS NOT in `collapsableLoops`, split `indicies` and don't add
+  // `index` to either.
+  //
+  // (2) If `index` IS in `collapsableLoops` but `consumerCollapseMap` maps
+  // `index` to a different collapsed loop then the other indicies,  split
+  // `indicies` and insert `index` into the new one.
+  //
+  // For example:
+  // producer reassociation = [[0, 1], [2, 3]]
+  // consumer reassociation = [0, 1, 2, 3]
+  // then, consumer reassociation gets updated to [[0, 1], [2, 3]] because
+  // [0, 1] and [2, 3] get collapsed into different loops
+  //
+  // (3) Otherwise, keep the index
+  constexpr long kUninitialized = -1;
+  SmallVector<ReassociationIndices> newReassociation;
+  for (ReassociationIndicesRef indicies : reassociation) {
+    // Track the loop index that `indicies` get collapsed into.
+    long collapseIntoIdx = kUninitialized;
+
+    // Holds dimensions that should be collapsed together
+    ReassociationIndices newIndicies;
+    for (int64_t index : indicies) {
+      // (1) Because `index` isn't collapsable, the indicies in `newIndicies`
+      // are no longer adjacent to the upcoming indicies. If there is >1 index
+      // to collapse, add it to the new reassociation. Otherwise, discard it
+      // because there is no dimension to collapse with.
+      if (!collapsableLoops.contains(index)) {
+        if (newIndicies.size() > 1) {
+          newReassociation.push_back(std::move(newIndicies));
+        }
+        newIndicies.clear();
+        collapseIntoIdx = kUninitialized;
+        continue;
+      } else if (collapseIntoIdx == kUninitialized) {
+        collapseIntoIdx = consumerCollapseMap.at(index);
+      } else if (consumerCollapseMap.at(index) != collapseIntoIdx) {
```

Comment:
Correct me if I'm wrong. I think the indices in the ReassociationIndices should be all different. To be more specific, they are just a sequence from 0 to rank-1 if we flatten it. So this can just be `else`?

```suggestion
      } else {
```

Reply:
Okay I added some comments and also tried to clean up the if statement logic a bit. That was meant to handle the case where `index` is collapsible but current loops.

Reply:
e.g. see the added testcase where there is an elementwise -> generic (2 parallel + 2 reduction).